### DISCUSSION TITLE
Importing XS data

### DISCUSSION THREAD

**Discussion by VikramGovindarajan** [seamplex/feenox#16] [Q&A]
In the example "IAEA 3D PWR Benchmark", the two group constants (cross sections, diffusion coefficients, etc.) are directly hard coded in the main input file itself. Is it possible to import this data from standard formatted files?

**Comment by gtheler** [comment]
Sure. How do these standard-formatted files look like?

**Reply by VikramGovindarajan** [reply]
The format can be defined in a separate file as tab-separated values (TSV) or comma-separated values (CSV) as shown below:

Material | D1 | D2 | ... | DN | Sigma_s1.2 | Sigma_s1.3 | .... | Sigma_s(N-1)(N) | Sigma_a1 | Sigma_a2 | .... | Sigma_aN | nuSigma_f1 | Sigma_f2 | .... | Sigma_fN | 
|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|--|
|Fuel1|||||||||||||||||
|Fuel2|||||||||||||||||

The advantage would be to avoid cluttering the main input file.

**Comment by gtheler** [comment]
I'll need to ask more "why" questions because I think I misunderstood you:

What's the reason you want to split the XS data into a separate file? I assumed you had complex dependencies of XSs with temperature, boron, etc. and you had those dependencies in point-wise-defined functions (a.k.a. "tables" which is a term I don't like at all). 
But if the reason is only not to clutter up the input file, the straightforward solution is to put the `MATERIAL` lines in a separate file and include it from the first. 

Prepare a file named `xs.fee`:

```
MATERIAL fuel1     D1=1.5 D2=0.4 Sigma_s1.2=0.02 Sigma_a1=0.01 Sigma_a2=0.08  nuSigma_f2=0.135
MATERIAL fuel2     D1=1.5 D2=0.4 Sigma_s1.2=0.02 Sigma_a1=0.01 Sigma_a2=0.085 nuSigma_f2=0.135
MATERIAL fuel2rod  D1=1.5 D2=0.4 Sigma_s1.2=0.02 Sigma_a1=0.01 Sigma_a2=0.13  nuSigma_f2=0.135
MATERIAL reflector D1=2.0 D2=0.3 Sigma_s1.2=0.04 Sigma_a1=0    Sigma_a2=0.01  nuSigma_f2=0
MATERIAL reflrod   D1=2.0 D2=0.3 Sigma_s1.2=0.04 Sigma_a1=0    Sigma_a2=0.055 nuSigma_f2=0
```

And then the main input would look like

```
PROBLEM neutron_diffusion 3D GROUPS 2
DEFAULT_ARGUMENT_VALUE 1 quarter
READ_MESH iaea-3dpwr-$1.msh
INCLUDE xs.fee
BC vacuum   vacuum=0.4692
BC mirror   mirror
SOLVE_PROBLEM
WRITE_RESULTS FORMAT vtk
```

If the reason is that you already have XS data in that format and you want to prepare an input for FeenoX using that data, I suggest you to write a small converter from the CSV to something akin to the `xs.fee` file above. Previous versions (i.e. milonga) had keywords like `READ` that would be able to read numerical data from ASCII file but I dropped them based on deep philosophical programming questions.

If you actually have "tables" of XSs vs. indepdendent variables (temperature, boron, etc.) then definitely the way to go is CSV but in a different format where each column holds one value (say one column for D1, one for D2, one for Sigma_a1, etc.) and each row one independent value (say one row for T=100, one for T=120, etc.)

Confirm which of the "why do you want that" is it so I can suggest the best solution.

**Comment by gtheler** [comment]
Any update @VikramGovindarajan ?

**Comment by jeremyonscale** [comment]
Nothing @VikramGovindarajan ?

**Comment by gtheler** [comment]
Pinging @VikramGovindarajan again just in case.
